torch/_refs/nn/functional/__init__.py

```diff
@@ -294,3 +295,37 @@ def tanhshrink(a: TensorLikeType) -> TensorLikeType:
             "Expected a tensor input for an elementwise unary operation!"
         )
     return refs.sub(a, refs.tanh(a))
+
+
+@elementwise_unary_scalar_wrapper
+@elementwise_type_promotion_wrapper(
+    type_promoting_args=("a"),
+    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+)
+def hardtanh(
+    a: TensorLikeType,
+    min_val: NumberType = -1,
+    max_val: NumberType = 1,
+    inplace: bool = False,
+) -> TensorLikeType:
+    """
+    Reference implementation of torch.nn.functional.hardtanh
+    """
+    if inplace:
+        raise NotImplementedError
+    if not isinstance(a, TensorLike):
+        raise RuntimeError(
+            "Expected a tensor input for an elementwise unary operation!"
+        )
+    if utils.is_boolean_dtype(a.dtype):
+        raise RuntimeError("Bool inputs not supported for hardtanh")
+
+    # preserve legacy behavior of boundaries not causing type promotion
+    if utils.is_integer_dtype(a.dtype):
+        min_val = int(min_val)
+        max_val = int(max_val)
+        if not (a.dtype != torch.uint8 or (min_val >= 0 and max_val >= 0)):
+            raise RuntimeError(
+                "Cannot do hardtanh on an unsigned type with negative limits"
+            )
+    return torch.clamp(a, min_val, max_val)
```

Comment:
should we call ref `clamp` here, or are you better off with `torch.clamp`? Ref clamp calls minimum/maximum

Reply:
I don’t want to force decomposition down to prims here (or in general).

Reply:
what's the general rule of thumb for when to call torch.X vs ref.X and how does it apply to this specific case?

Reply:
There's a little personal preference because not all the features for calling torch.foo are available at this time, so using torch calls in a ref will cause the meta test to fail. That said, here are some good rules:

- Once all functionality for calling torch.foo correctly is available, prefer calling torch.foo
- If torch.foo doesn't support something that refs.foo does (like scalar inputs), call refs.foo instead

Today we don't force torch.foo or refs.foo (it's easy to change this later) because some decomps want torch calls and other decomps are focused on validating metadata correctness, and we (temporarily) have to choose between the two.